tests/update/tests.py

```diff
@@ -346,3 +358,199 @@ def test_order_by_update_on_related_field(self):
         self.assertEqual(updated, 1)
         related.refresh_from_db()
         self.assertEqual(related.name, "new")
+
+
+class TransformUpdateTests(TestCase):
+    def test_get_update_expression_not_implemented(self):
+        date_created = datetime.datetime(3000, 1, 1, 21, 22, 23)
+        user_preferences = UserPreferences.objects.create(date_created=date_created)
+
+        with self.assertRaisesMessage(
+            NotImplementedError,
+            "Using ExtractYear is not supported in QuerySet.update().",
+        ):
+            UserPreferences.objects.update(date_created__year=2024)
+
+        user_preferences.refresh_from_db()
+        self.assertEqual(user_preferences.date_created, date_created)
+
+    def test_invalid_transform(self):
+        date_created = datetime.datetime(3000, 1, 1, 21, 22, 23)
+        user_preferences = UserPreferences.objects.create(date_created=date_created)
+
+        with self.assertRaisesMessage(
+            FieldError,
+            "foo is not a valid transform on DateTimeField.",
+        ):
+            UserPreferences.objects.update(date_created__foo=2024)
+
+        user_preferences.refresh_from_db()
+        self.assertEqual(user_preferences.date_created, date_created)
+
+
+@skipUnlessDBFeature("supports_partial_json_update")
+class JSONKeyTransformUpdateTests(TestCase):
+    def test_update_json_key_transform(self):
+        user_preferences = UserPreferences.objects.create(
+            settings={
+                "theme": {"color": "black", "font": "Arial"},
+                "notifications": {"email": False, "sms": True},
+            }
+        )
+        UserPreferences.objects.update(settings__theme__color="white")
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+        self.assertEqual(
+            user_preferences.settings,
+            {
+                "theme": {"color": "white", "font": "Arial"},
+                "notifications": {"email": False, "sms": True},
+            },
+        )
+
```

Comment:
So I was curious what happens when we do `update_or_create` using this syntax and the answer is, it doesn't fail but it doesn't update (the following test fails expecting it would update).
```suggestion

    def test_update_or_create_json_key_transform(self):
        user_preferences = UserPreferences.objects.create(
            settings={
                "theme": {"color": "black", "font": "Arial"},
                "notifications": {"email": False, "sms": True},
            }
        )
        updated_user_preferences, created = UserPreferences.objects.update_or_create(
            defaults={"settings__theme__color": "white"},
            id=user_preferences.id
        )
        self.assertEqual(created, False)
        self.assertEqual(
            updated_user_preferences.settings,
            {
                "theme": {"color": "white", "font": "Arial"},
                "notifications": {"email": False, "sms": True},
            },
        )

```

Can you investigate whether adding support for `update_or_create` feasible?

Reply:
Thanks Sarah, that's a good point! We haven't investigated into adding transforms support for `update_or_create`, but thanks to your review we looked into how `JSONSet` and `JSONRemove` behaves with `update_or_create()`.

We're going to open a separate PR for adding those two functions. In the meantime here are our findings for using `update_or_create` with the functions.

With the following example:

```py
updated_user_preferences, created = UserPreferences.objects.update_or_create(
    defaults={"settings": JSONSet("settings", theme__color="white")},
    id=user_preferences.id,
)
```

- In the case where it does an update (`created is False`), it worked well – the test passed.
- In the case where it does an insert (`created is True`), it fails because we'd be referring to the column for the row which hasn't been inserted yet. It makes sense that this would fail, but I think it would be nice if `JSONSet()` and `JSONRemove()` can fall back to an empty JSON object as the starting point in the insert case.
  - The main blocker we found in this case is that the `for_save` parameter in [`BaseExpression.resolve_expression()`](https://docs.djangoproject.com/en/stable/ref/models/expressions/#django.db.models.Expression.resolve_expression) doesn't let you differentiate whether it's an `INSERT` or an `UPDATE` query – it's set to `True` in both cases.
  - We hacked the places where we pass `for_save=True` and changed it to pass the strings `"create"` and `"update"` depending on the query, and we were able to make `JSONSet()` and `JSONRemove()` resolve the expression accordingly. See https://github.com/adzshaf/django/commit/207598f41e845b9f1db8dcdb3dc1f2850d009eeb for more details.
  - I'm not sure the best course of action on that, though. That being said, I just noticed [this forum thread](https://forum.djangoproject.com/t/add-support-for-atomic-upserts/35813/5) about atomic upserts that would also benefit from the ability to differentiate between `INSERT` vs `UPDATE` in `resolve_expression()`. I'll ask there and see if others have an opinion.